File: core/src/passes/dead_store.cpp
// core/src/passes/dead_store.cpp — F1-PS-008.
//
// A StoreMem whose value is overwritten by a subsequent StoreMem to the
// same (addr_ref, size) pair — with NO intervening memory operation
// that could observe it — is dead. We drop the earlier one.
//
// Implementation:
//   * Two-pass: first scan forward to collect dead-store indices,
//     second pass filters them out. Collecting indices first avoids
//     the fiddly "update pending-store iterator mid-emit" dance and
//     keeps the rewrite trivially linear.
//   * `pending_store[(addr, size)] = last_store_stmt_idx`. On a new
//     StoreMem to the same `(addr, size)`, the previous entry becomes
//     a dead-store index and is replaced.
//   * Any LoadMem / LoadMemTSO / StoreMemTSO clears `pending_store`
//     entirely — the pending store might have been observed.
//   * StoreMem to a DIFFERENT `(addr, size)` under this MVP simply
//     adds itself to the table without clearing — different keys don't
//     alias each other in the address-ref space. (This is the
//     soundness-relevant invariant: two StoreMems with distinct addr
//     refs are guaranteed by SSA to write different concrete addresses
//     within a single call path, but not across all schedules. Still
//     safe per-block since we don't cross block boundaries.)
//
// Scope decisions:
//   * StoreMemTSO is release-ordered — another thread may observe it
//     even in a block that appears to overwrite it. Never killed.
//   * `addr_ref` equality, not value equality: %a == %a counts, but
//     two distinct LoadReg results pointing at the same guest reg
//     don't. Missing some opportunities is fine; soundness first.

#include "prisma/passes.hpp"

#include <cstddef>
#include <unordered_map>
#include <unordered_set>
#include <variant>

namespace prisma::passes {

namespace {

struct StoreKey {
    ir::Ref    addr;
    ir::OpSize size;
    bool operator==(const StoreKey&) const = default;
};

struct StoreKeyHash {
    std::size_t operator()(const StoreKey& k) const noexcept {
        return std::hash<ir::Ref>{}(k.addr) * 31u
             + static_cast<std::size_t>(k.size);
    }
};

}  // namespace

std::vector<ir::Stmt>
dead_store_eliminate(const std::vector<ir::Stmt>& stmts) {
    std::unordered_map<StoreKey, std::size_t, StoreKeyHash> pending;
    std::unordered_set<std::size_t> dead_indices;

    for (std::size_t i = 0; i < stmts.size(); ++i) {
        const auto& s = stmts[i];
        if (std::holds_alternative<ir::StoreMem>(s.op)) {
            const auto& st = std::get<ir::StoreMem>(s.op);
            const StoreKey k{st.addr, st.size};
            auto it = pending.find(k);
            if (it != pending.end()) {
                dead_indices.insert(it->second);
                it->second = i;
            } else {
                pending[k] = i;
            }
        }
        else if (std::holds_alternative<ir::LoadMem>(s.op)
              || std::holds_alternative<ir::LoadMemTSO>(s.op)
              || std::holds_alternative<ir::StoreMemTSO>(s.op)
              || std::holds_alternative<ir::Fence>(s.op)) {
            // Could observe / synchronise with a pending store.
            pending.clear();
        }
        // Pure ops don't interact with memory — table survives.
    }

    // Filter.
    std::vector<ir::Stmt> out;
    out.reserve(stmts.size() - dead_indices.size());
    for (std::size_t i = 0; i < stmts.size(); ++i) {
        if (!dead_indices.contains(i)) out.push_back(stmts[i]);
    }
    return out;
}

}  // namespace prisma::passes

File: shell/prisma-passes/src/dead_store.rs
//! Dead store elimination pass.
//!
//! Mirrors C++ `dead_store_eliminate`. A `StoreMem` to `(addr_ref, size)`
//! that is overwritten by a later `StoreMem` to the same key, with no
//! intervening memory op that could observe it, is dropped. `StoreMemTSO`
//! (release-ordered), `LoadMem`, `LoadMemTSO`, and `Fence` clear the pending
//! table — the earlier store may have been observed.

use std::collections::{HashMap, HashSet};

use prisma_ir::{BasicBlock, Function, Op, OpSize, Ref, Stmt};

use crate::Pass;

#[derive(Debug, Default, Clone, Copy, PartialEq, Eq)]
pub struct DeadStore;

impl DeadStore {
    #[must_use]
    pub const fn new() -> Self {
        Self
    }
}

impl Pass for DeadStore {
    fn name(&self) -> &'static str {
        "dead_store_eliminate"
    }

    fn run(&self, func: Function) -> Function {
        dead_store_eliminate(func)
    }
}

/// Per-block dead store elimination.
pub fn dead_store_eliminate(func: Function) -> Function {
    let blocks = func
        .blocks
        .into_iter()
        .map(|block| {
            let mut pending: HashMap<(Ref, OpSize), usize> = HashMap::new();
            let mut dead: HashSet<usize> = HashSet::new();

            for (i, stmt) in block.stmts.iter().enumerate() {
                match &stmt.op {
                    Op::StoreMem(st) => {
                        let key = (st.addr, st.size);
                        if let Some(prev) = pending.insert(key, i) {
                            dead.insert(prev);
                        }
                    }
                    Op::LoadMem(_) | Op::LoadMemTSO(_) | Op::StoreMemTSO(_) | Op::Fence(_) => {
                        pending.clear();
                    }
                    _ => {}
                }
            }

            let stmts = block
                .stmts
                .into_iter()
                .enumerate()
                .filter_map(|(i, s)| if dead.contains(&i) { None } else { Some(s) })
                .collect();

            BasicBlock { id: block.id, stmts }
        })
        .collect();

    Function {
        blocks,
        entry: func.entry,
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use prisma_ir::{Fence, FenceKind, LoadMem, StoreMem, StoreMemTSO};

    fn store(addr: u32, value: u32, size: OpSize) -> Stmt {
        Stmt::new(None, Op::StoreMem(StoreMem { addr, value, size }))
    }

    #[test]
    fn overwritten_store_is_dropped() {
        let func = Function {
            entry: 0,
            blocks: vec![BasicBlock {
                id: 0,
                stmts: vec![store(5, 0, OpSize::I64), store(5, 1, OpSize::I64)],
            }],
        };
        let out = dead_store_eliminate(func);
        assert_eq!(out.blocks[0].stmts.len(), 1);
        match &out.blocks[0].stmts[0].op {
            Op::StoreMem(s) => assert_eq!(s.value, 1),
            other => panic!("unexpected {other:?}"),
        }
    }

    #[test]
    fn intervening_load_keeps_both() {
        let func = Function {
            entry: 0,
            blocks: vec![BasicBlock {
                id: 0,
                stmts: vec![
                    store(5, 0, OpSize::I64),
                    Stmt::new(Some(9), Op::LoadMem(LoadMem { addr: 6, size: OpSize::I64 })),
                    store(5, 1, OpSize::I64),
                ],
            }],
        };
        let out = dead_store_eliminate(func.clone());
        assert_eq!(out, func);
    }

    #[test]
    fn tso_store_never_killed() {
        let func = Function {
            entry: 0,
            blocks: vec![BasicBlock {
                id: 0,
                stmts: vec![
                    Stmt::new(None, Op::StoreMemTSO(StoreMemTSO { addr: 5, value: 0, size: OpSize::I64 })),
                    store(5, 1, OpSize::I64),
                ],
            }],
        };
        let out = dead_store_eliminate(func.clone());
        assert_eq!(out, func);
    }

    #[test]
    fn different_addr_keeps_both() {
        let func = Function {
            entry: 0,
            blocks: vec![BasicBlock {
                id: 0,
                stmts: vec![store(5, 0, OpSize::I64), store(6, 1, OpSize::I64)],
            }],
        };
        let out = dead_store_eliminate(func.clone());
        assert_eq!(out, func);
    }

    #[test]
    fn fence_blocks_elimination() {
        let func = Function {
            entry: 0,
            blocks: vec![BasicBlock {
                id: 0,
                stmts: vec![
                    store(5, 0, OpSize::I64),
                    Stmt::new(None, Op::Fence(Fence { kind: FenceKind::Sfence })),
                    store(5, 1, OpSize::I64),
                ],
            }],
        };
        let out = dead_store_eliminate(func.clone());
        assert_eq!(out, func);
    }
}
